src/tools/ScenarioMeasurement/Startup/Crossgen2Parser.cs

```diff
@@ -0,0 +1,122 @@
+﻿using Microsoft.Diagnostics.Tracing.Parsers;
+using Reporting;
+using System;
+using System.Collections.Generic;
+
+namespace ScenarioMeasurement
+{
+    public class Crossgen2Parser : IParser
+    {
+        public static readonly string ProviderName = "Microsoft-ILCompiler-Perf";
+        public void EnableKernelProvider(ITraceSession kernel)
+        {
+            kernel.EnableKernelProvider(TraceSessionManager.KernelKeyword.Process, TraceSessionManager.KernelKeyword.Thread, TraceSessionManager.KernelKeyword.ContextSwitch);
+        }
+
+        public void EnableUserProviders(ITraceSession user)
+        {
+            user.EnableUserProvider(ProviderName);
+        }
+
+        public IEnumerable<Counter> Parse(string mergeTraceFile, string processName, IList<int> pids, string commandLine)
+        {
+            var loadingParser = new EventParser("Loading");
+            var emittingParser = new EventParser("Emitting");
+            var jittingParser = new EventParser("Jit");
+            var compilationParser = new EventParser("Compilation");
+
+
+            using (var source = new TraceSourceManager(mergeTraceFile))
+            {
+                var dynamicParser = new DynamicTraceEventParser(source.Source);
+
+                loadingParser.AddStartStopCallbacks(source, dynamicParser, ProviderName, pids);
+                emittingParser.AddStartStopCallbacks(source, dynamicParser, ProviderName, pids);
+                jittingParser.AddStartStopCallbacks(source, dynamicParser, ProviderName, pids);
+                compilationParser.AddStartStopCallbacks(source, dynamicParser, ProviderName, pids);
+
+                source.Process();
+            }
+
+            var processTimeParser = new ProcessTimeParser();
+            Counter processTimeCounter = null;
+            foreach (var counter in processTimeParser.Parse(mergeTraceFile, processName, pids, commandLine))
+            {
+                if(counter.Name == "Process Time"){
+                    processTimeCounter = counter;
+                }
+            }
+
+            return new[] {
+                processTimeCounter,
+                new Counter() { Name = "Loading Interval", MetricName = "ms", DefaultCounter=false, TopCounter=true, Results = loadingParser.Intervals.ToArray() },
+                new Counter() { Name = "Emitting Interval", MetricName = "ms", DefaultCounter=false, TopCounter=true, Results = emittingParser.Intervals.ToArray() },
+                new Counter() { Name = "Jit Interval", MetricName = "ms", DefaultCounter=false, TopCounter=true, Results = jittingParser.Intervals.ToArray() },
+                new Counter() { Name = "Compilation Interval", MetricName = "ms", DefaultCounter=false, TopCounter=true, Results = compilationParser.Intervals.ToArray() }
+            };
+        }
+    }
+
+    public sealed class EventParser
```

Comment:
Should this go into some shared spot? Looks pretty generic.

Reply:
This class only finds a pair of start/stop events with the specified event name. Its matching algorithm is designed for crossgen2 events, which is a bit loose compared to the matching algorithm of kernel events. It can definitely go to a shared spot if it could apply to other parsers.